src/main/java/com/datadoghq/sketch/ddsketch/store/PaginatedStore.java

```diff
@@ -0,0 +1,387 @@
+            canShiftRight = null == pages[pages.length - i - 1];
+        }
+        if (canShiftRight) {
+            System.arraycopy(pages, 0, pages, requiredExtension, pages.length - requiredExtension);
+        } else {
+            double[][] newPages = new double[pages.length + aligned(requiredExtension)][];
+            System.arraycopy(pages, 0, newPages, requiredExtension, pages.length);
+            this.pages = newPages;
+        }
+        Arrays.fill(pages, 0, requiredExtension, null);
+        this.minPageIndex = pageIndex;
+    }
+
+    private void extendTo(int pageIndex) {
+        this.pages = Arrays.copyOf(pages, pages.length + aligned(pageIndex - minPageIndex + 1));
+    }
+
+    @Override
+    public void mergeWith(Store store) {
+        if (store.isEmpty()) {
+            return;
+        }
+        if (store instanceof PaginatedStore) {
+            mergeWith((PaginatedStore) store);
+        } else {
+            store.getStream().forEach(this::add);
+        }
+    }
+
+    private void mergeWith(PaginatedStore store) {
+        if (isEmpty()) {
+            this.pages = deepCopy(store.pages);
+            this.minPageIndex = store.minPageIndex;
+        } else {
+            int min = minPageIndex;
+            int max = minPageIndex + pages.length;
+            int storeMin = store.minPageIndex;
+            int storeMax = store.minPageIndex + store.pages.length;
+            // no overlap cases
+            if (max < storeMin) {
+                extendTo(storeMax);
+                for (int i = 0; i < store.pages.length; ++i) {
+                    double[] page = store.pages[i];
+                    if (null != page) {
+                        pages[i + storeMin - min] = Arrays.copyOf(page, page.length);
+                    }
+                }
+            } else if (min > storeMax) {
+                shiftPagesRight(storeMin);
+                for (int i = 0; i < store.pages.length; ++i) {
+                    double[] page = store.pages[i];
+                    if (null != page) {
+                        pages[i] = Arrays.copyOf(page, page.length);
+                    }
+                }
+            } else if (min < storeMin) {
+                if (storeMax > max) {
+                    extendTo(storeMax);
+                }
+                for (int i = 0; i < store.pages.length; ++i) {
+                    double[] page = store.pages[i];
+                    if (null != page) {
+                        double[] target = pages[i + storeMin - min];
+                        if (null == target) {
+                            pages[i + storeMin - min] = Arrays.copyOf(page, page.length);
+                        } else {
+                            for (int j = 0; j < page.length; ++j) {
+                                target[j] += page[j];
+                            }
+                        }
+                    }
+                }
+            } else {
+                shiftPagesRight(storeMin);
```

Comment:
So if `min == storeMin` then it looks like this would try to do a `0` length `arraycopy`.

Reply:
Yes, I missed this.